src/autopas/containers/octree/Octree.h

```diff
@@ -0,0 +1,240 @@
+
+  /**
+   * This particle container contains two cells. Both cells are octrees.
+   * - this->_cells[CellTypes::OWNED] yields the octree that contains all owned particles
+   * - this->_cells[CellTypes::HALO] yields the octree that contains all halo particles
+   */
+  enum CellTypes { OWNED = 0, HALO = 1 };
+
+  /**
+   * Construct a new octree with two sub-octrees: One for the owned particles and one for the halo particles.
+   * @param boxMin The minimum coordinate of the enclosing box
+   * @param boxMax The maximum coordinate of the enclosing box
+   * @param cutoff The cutoff radius
+   * @param skin The skin radius
+   * @param cellSizeFactor The cell size factor
+   */
+  Octree(std::array<double, 3> boxMin, std::array<double, 3> boxMax, const double cutoff, const double skin,
+         const double cellSizeFactor)
+      : CellBasedParticleContainer<ParticleCell>(boxMin, boxMax, cutoff, skin) {
+    // TODO(johannes): Obtain this from a configuration, reported in https://github.com/AutoPas/AutoPas/issues/624
+    int unsigned treeSplitThreshold = 16;
+
+    double interactionLength = this->getInteractionLength();
+    this->_cells.push_back(
+        OctreeNodeWrapper<Particle>(boxMin, boxMax, treeSplitThreshold, interactionLength, cellSizeFactor));
+
+    // Extend the halo region with cutoff + skin in all dimensions
+    auto haloBoxMin = utils::ArrayMath::subScalar(boxMin, this->getInteractionLength());
+    auto haloBoxMax = utils::ArrayMath::addScalar(boxMax, this->getInteractionLength());
+    this->_cells.push_back(
+        OctreeNodeWrapper<Particle>(haloBoxMin, haloBoxMax, treeSplitThreshold, interactionLength, cellSizeFactor));
+  }
+
+  [[nodiscard]] std::vector<ParticleType> updateContainer() override {
+    // This is a very primitive and inefficient way to recreate the container:
+    // 1. Copy all particles out of the container
+    // 2. Clear the container
+    // 3. Insert the particles back into the container
+
+    // leaving: all outside boxMin/Max
+
+    // TODO(johannes): Make this less indirect. (Find a better way to iterate all particles inside the octree to change
+    //  this function back to a function that actually copies all particles out of the octree.)
+    //  The problem is captured by https://github.com/AutoPas/AutoPas/issues/622
+    std::vector<Particle *> particleRefs;
+    this->_cells[CellTypes::OWNED].appendAllParticles(particleRefs);
+    std::vector<Particle> particles;
+    auto result = std::vector<ParticleType>();
+    for (auto *p : particleRefs) {
+      if (utils::inBox(p->getR(), this->getBoxMin(), this->getBoxMax())) {
+        particles.push_back(*p);
+      } else {
+        result.push_back(*p);
+      }
+    }
+    this->deleteAllParticles();
+
+    for (auto &particle : particles) {
+      addParticleImpl(particle);
+    }
+
+    // This should happen according to the documentation in the ParticleContainerInterface#updateContainer
+    deleteHaloParticles();
+
```

Comment:
true but calling this only makes sense if the `addParticleImpl()` call might have added any halo particles. Otherwise the call to `deleteAllParticles()` some lines above should already have taken care of the halos

Reply:
You are right, I remove the duplication in c51cfb6.